rp2040-hal/src/dma/single_channel.rs

```diff
@@ -0,0 +1,208 @@
+    fn ch(&self) -> &rp2040_pac::dma::CH {
+        self.regs()
+    }
+
+    fn id(&self) -> u8 {
+        CH::id()
+    }
+}
+
+impl<CH1: ChannelIndex, CH2: ChannelIndex> SingleChannel for (Channel<CH1>, Channel<CH2>) {
+    fn ch(&self) -> &rp2040_pac::dma::CH {
+        self.0.regs()
+    }
+
+    fn id(&self) -> u8 {
+        CH1::id()
+    }
+}
+
+pub(crate) trait ChannelConfig {
+    fn config<WORD, FROM, TO>(
+        &mut self,
+        from: &FROM,
+        to: &mut TO,
+        pace: Pace,
+        chain_to: Option<u8>,
+        start: bool,
+    ) where
+        FROM: ReadTarget<ReceivedWord = WORD>,
+        TO: WriteTarget<TransmittedWord = WORD>;
+
+    fn set_chain_to_enabled<CH: SingleChannel>(&mut self, other: &mut CH);
+    fn start(&mut self);
+}
+
+impl<CH: SingleChannel> ChannelConfig for CH {
+    fn config<WORD, FROM, TO>(
+        &mut self,
+        from: &FROM,
+        to: &mut TO,
+        pace: Pace,
+        chain_to: Option<u8>,
+        start: bool,
+    ) where
+        FROM: ReadTarget<ReceivedWord = WORD>,
+        TO: WriteTarget<TransmittedWord = WORD>,
+    {
+        // Configure the DMA channel.
+        assert!(
+            mem::size_of::<WORD>() != 8,
+            "DMA does not support transferring 64bit data"
+        );
+        let (src, src_count) = from.rx_address_count();
+        let src_incr = from.rx_increment();
+        let (dest, dest_count) = to.tx_address_count();
+        let dest_incr = to.tx_increment();
+        const TREQ_UNPACED: u8 = 0x3f;
+        let treq = match pace {
+            Pace::PreferSource => FROM::rx_treq().or_else(TO::tx_treq).unwrap_or(TREQ_UNPACED),
+            Pace::PreferSink => TO::tx_treq().or_else(FROM::rx_treq).unwrap_or(TREQ_UNPACED),
+        };
+        let len = u32::min(src_count, dest_count);
+        self.ch().ch_al1_ctrl.write(|w| unsafe {
+            w.data_size().bits(mem::size_of::<WORD>() as u8 >> 1);
+            w.incr_read().bit(src_incr);
+            w.incr_write().bit(dest_incr);
+            w.treq_sel().bits(treq);
+            w.chain_to().bits(chain_to.unwrap_or_else(|| self.id()));
+            w.en().bit(true);
+            w
+        });
+        self.ch().ch_read_addr.write(|w| unsafe { w.bits(src) });
+        self.ch().ch_trans_count.write(|w| unsafe { w.bits(len) });
+        if start {
+            self.ch()
+                .ch_al2_write_addr_trig
+                .write(|w| unsafe { w.bits(dest) });
+        } else {
+            self.ch().ch_write_addr.write(|w| unsafe { w.bits(dest) });
+        }
+    }
+
+    fn set_chain_to_enabled<CH2: SingleChannel>(&mut self, other: &mut CH2) {
+        // We temporarily pause the channel when setting CHAIN_TO, to prevent any race condition
+        // that could occur, as we need to check afterwards whether the channel was successfully
+        // chained to this channel or whether this channel was already completed. If we did not
```

Comment:
How does that interact with DREQ grants? Are they still counted while the channel is disabled?

If they are note, the DREQ counter could get out of sync.

Reply:
What do you mean by "DREQ grants"? I can't find this term anywhere.

Reply:
I did some experiments, temporarily pausing a channel and reading DBG_CTDREQ. Looks like the DREQ mechanism keeps counting while a the `en` flag is cleared, so this shoud be fine!

Reply:
Ah, good. I read that section of the datasheet + the pico-sdk and it seemed like it would just keep counting, but I was going to test it out to be sure. Now that you've tested it I'll trust that it works. :+1: